src/fastpath/fastpath.c

```diff
@@ -15,6 +15,268 @@
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_ARCH_ARM
+static inline
+#ifndef CONFIG_ARCH_ARM_V6
+FORCE_INLINE
+#endif
+#endif
+void NORETURN fastpath_signal(word_t cptr, word_t msgInfo)
+{
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    pde_t stored_hw_asid;
+    word_t fault_type;
+    dom_t dom;
+
+    /* Get fault type. */
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
+
+    /* We land up in here on every SYSCALL_SEND, so we need to check that we are
+     * actually signalling a notification */
+
+    /* Lookup the cap */
+    cap_t cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, cptr);
```

Comment:
Possibly move this to top to avoid pipeline stalls after line 33.

Reply:
https://sjmulder.nl/dl/pdf/unsorted/2012%20-%20Blackham%20&%20Heiser%20-%20Correct,%20Fast,%20Maintainable,%20Choose%20Any%20Three.pdf

Reply:
Super interesting paper :-). I've lifted the cap lookup so we now do it immediately getting the fault type.

Reply:
Let me know if this is okay to mark as resolved.